src/components/graph/ExerciseGraphLayout.ts

```diff
@@ -0,0 +1,337 @@
+
+    console.log(this)
+  }
+
+  /**
+   * This private method checks the current graph if it contains a cycle.  If a cycle exists:
+   *   - This graph cannot be rendered
+   *   - Points at a larger problem with the exercise concept pre-requisites
+   */
+  private containsCycle(): boolean {
+    const visited: Array<boolean> = new Array(this.adjacency.length).fill(false)
+    const isOnStack: Array<boolean> = new Array(this.adjacency.length).fill(
+      false
+    )
+    const stack: Array<number> = []
+
+    // Attempt to start from each node in the case the graph is disjunct
+    for (let source = 0; source < this.adjacency.length; source += 1) {
+      if (visited[source]) continue
+      stack.push(source)
+
+      // loop until the stack is empty (i.e. you have visited all of the nodes)
+      while (stack.length > 0) {
+        const n = stack[stack.length - 1]
+
+        if (!visited[n]) {
+          visited[n] = true
+          isOnStack[n] = true
+        } else {
+          isOnStack[n] = false
+          stack.pop()
+        }
+
+        for (const next of this.adjacency[n]) {
+          if (!visited[next]) {
+            stack.push(next)
+          } else if (isOnStack[next]) {
+            console.error(
+              `Graph Cycle Detected '${
+                this.graph.lookupByIndex.get(n)?.slug
+              }' to '${this.graph.lookupByIndex.get(next)?.slug}'`
+            )
+            return true
+          }
+        }
+      }
+    }
+    return false
+  }
+
+  /**
+   * computeNodeDepths
+   * This computes the depth of all nodes, first exercises then prerequisite concepts
+   * that are not yet represented in an exercise.
+   */
+  private computeNodeDepths(): void {
+    const inDegrees = this.findIndegreeOfExercises()
+    const sourceExercises = this.findSourceExercises(inDegrees)
+    const exerciseDepths = this.findExerciseDepths(sourceExercises)
+    const maximumDepth = Math.max.apply(null, exerciseDepths)
+    this.exerciseLayers = this.constructExerciseLayers(
+      exerciseDepths,
+      maximumDepth
+    )
+    this.conceptLayers = this.constructConceptLayers(this.exerciseLayers)
+  }
+
+  /**
+   * Use the adjacency matrix to find the in-degree of each
+   * node (the number of edges pointing to the node)
+   *
+   * Return an array of numbers, where each index (representing
+   * an ExerciseGraphNode) holds the indegree value
+   */
+  private findIndegreeOfExercises(): Array<number> {
+    const indegrees = new Array(this.adjacency.length).fill(0)
+    for (const tos of this.adjacency) {
+      for (const to of tos) {
+        indegrees[to] += 1
+      }
+    }
+    return indegrees
+  }
+
+  /**
+   * Nodes with inDegree of 0 are source nodes (roots) of the graph
+   */
+  private findSourceExercises(inDegrees: Array<number>): Array<number> {
+    const sources: Array<number> = []
+    inDegrees.forEach((inDegree, i) => {
+      if (inDegree === 0) sources.push(i)
+    })
+    return sources
```

Comment:
```suggestion
    return inDegrees.map((inDegree, i) => inDegree === 0 ? i : undefined)
                    .filter(Boolean)
```

I would probably do this, or use `reduce`:

```suggestion
    return inDegrees.reduce((result, inDegree, i) => {
      if (inDegree === 0) {
        result.push(i)
      }
      
      return result
    }, [] as number[])
```
Or alternatively:


```suggestion
    return inDegrees.reduce((result, inDegree, i) => {
      return inDegree === 0 ? [...result, i] : result
    }, [] as number[])
```

Yours works just fine, I just don't like the bookkeeping (create variable, then iterate with conditional push, then return that variable).

If you keep yours, make sure you don't inline ifs:

```suggestion
    const sources: Array<number> = []
    inDegrees.forEach((inDegree, i) => {
      if (inDegree === 0) {
        sources.push(i)
      }
    })
    return sources
```